samples/Rewrite/Structure2/UrlRewriteMiddleware.cs

```diff
@@ -0,0 +1,136 @@
+﻿using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Http;
+using Rewrite.ConditionParser;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text.RegularExpressions;
+using System.Threading.Tasks;
+
+namespace Rewrite.Structure2
+{
+    public class UrlRewriteMiddleware
+    {
+        private readonly RequestDelegate _next;
+        private readonly UrlRewriteOptions _options;
+        public UrlRewriteMiddleware(RequestDelegate next, UrlRewriteOptions options)
+        {
+            if (next == null)
+            {
+                throw new ArgumentNullException(nameof(next));
+            }
+
+            if (options == null)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
+
+            _next = next;
+            _options = options;
+        }
+
+        public async Task Invoke(HttpContext context)
+        {
+
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+            foreach (Rule rule in _options.Rules)
+            {
+                // 1. Check if the certain component of the path matches with the initial match.
+                // TODO logic for flags.
+                // TODO dont initialize the regex here for every request, have it initialized in 
+                //      either the parsing on lazily initialize and save here
+                Regex compare = new Regex(rule.InitialRule.Variable);
+                Match results = compare.Match(context.Request.Path.ToString());
+                // If the match came back negative (the rule xor'd with whether or not it is inverted)
+                // go to the next rule.
+                if (!(results.Success ^ rule.InitialRule.Invert))
+                {
+                    continue;
+                }
+                // 2. Go through all conditions and compare them to the created string
+                Match previous = Match.Empty;
+                var i = 0;
+                for (i = 0; i < rule.Conditions.Count; i++)
+                {
+                    Condition condition = rule.Conditions[i];
+                    string concatTestString = CollectTestStringFromContext(condition.TestStringSegments, context, previous, results);
+                    bool pass = false;
+                    switch (condition.ConditionRegex.Type)
+                    {
+                        case ConditionType.FileTest:
+                            pass = CheckFileCondition(concatTestString, condition, context);
```

Comment:
I would move all these in a `RuleEvaluator` class that part of the rewriter project. Then the middleware is `dumb` and just calls in that